modules/gateway/nodes.go

```diff
@@ -33,24 +35,32 @@ func (g *Gateway) addNode(addr modules.NetAddress) error {
 
 // pingNode verifies that there is a reachable node at the provided address
 // by performing the Sia gateway handshake protocol.
-func (g *Gateway) pingNode(addr modules.NetAddress) error {
+func (g *Gateway) pingNode(addr modules.NetAddress) (err error) {
 	// Ping the untrusted node to see whether or not there's actually a
 	// reachable node at the provided address.
 	conn, err := g.dial(addr)
 	if err != nil {
-		return err
+		return
 	}
 	defer conn.Close()
-	// If connection succeeds, supply an unacceptable version so that we
-	// will not be added as a peer.
-	//
-	// NOTE: this is a somewhat clunky way of specifying that you didn't
-	// actually want a connection.
-	_, err = connectVersionHandshake(conn, "0.0.0")
-	if err == errPeerRejectedConn {
-		err = nil // we expect this error
+
+	// Read the node's version.
+	remoteVersion, err := connectVersionHandshake(conn, build.Version)
+	if err != nil {
+		return
 	}
-	return err
+
+	if build.VersionCmp(remoteVersion, sessionHandshakeUpgradeVersion) < 0 {
+		return // for older versions, this is where pinging ends
```

Comment:
At first I thought we could mitigate this by checking the remote version and writing `0.0.0` if they were below the upgrade version. But that's not possible because we send our version first and then read the remote version.

Regardless, I don't think this will cause any serious issues. The old peer will perceive it as a connection error. Worst case, it adds a line to the old peer's gateway.log, but I think that's only enabled on debug builds.